Dev/async refactor

RandoomWalks opened the pull request:
Refactor Kademlia implementation into modular structure and update time handling in Kademlia implementation

 - Replaced blocking I/O operations with asynchronous counterparts using Tokio, improving scalability and performance.
 - Changed time tracking within k-buckets from `SystemTime` to `Instant` to better support the async runtime.
 - Added serialization and deserialization for `Instant` to maintain compatibility with the existing data format.
 - Introduced an asynchronous main function with `#[tokio::main]` attribute to support async/await syntax.
 - Implemented a graceful shutdown mechanism using a message-passing channel.
 - Added periodic bucket refresh logic to maintain up-to-date network information.
 - Updated routing table and node handling to work within the async context.

RandoomWalks committed:
Move to main(compile OK)

RandoomWalks committed:
Refactor KademliaNode and RoutingTable for improved logging and error handling

 - Enhanced logging across various methods in KademliaNode and RoutingTable, providing clearer information on actions performed, such as node update
 message handling, and bootstrap process.
 - Improved error handling in message serialization and deserialization, with added error logs for potential failures.
 - Streamlined use statements by removing unused imports and organizing them more logically.
 - Simplified node eviction logic in KBucket with clearer scope separation and added logging for eviction actions.
 - Removed commented-out code for a cleaner codebase.
 - Added env_logger initialization in the main function for proper logger setup.
 - Minor code formatting for better readability, including consistent spacing and struct alignment.

RandoomWalks committed:
Refactor Kademlia implementation into modular structure

 This commit overhauls the Kademlia DHT node implementation to adopt a more modular structure. The changes include:

 - Extracting the `KBucket` logic into its own file `kbucket.rs`, encapsulating the k-bucket functionality and
 improving code maintainability.
 - Moving the main node logic to `node.rs`, which now contains the `KademliaNode` struct and associated methods for
 node operations such as bootstrapping, message handling, and running the main event loop.
 - Creating `routing_table.rs` to handle the routing table logic separately, allowing for clearer separation of
 concerns and easier future enhancements to the routing mechanism.
 - Defining message types and results in `message.rs`, which simplifies the message handling process and makes the
 message format more explicit.
 - Introducing `utils.rs` to store common constants and utility functions, such as the `NodeId` struct and its
 associated methods, which are used across different modules.

RandoomWalks committed:
Update Rust dependencies and project structure

 - Remove unused dependencies related to anstyle and colorchoice
 - Downgrade env_logger from 0.11.5 to 0.10.2
 - Replace is_terminal_polyfill with is-terminal and update to version 0.4.12
 - Add termcolor dependency version 1.4.1
 - Remove utf8parse as it's no longer required
 - Update winapi-util to version 0.1.9
 - Reorganize Cargo.toml, moving lib section and updating dependencies
   - Specify versions for serde, bincode, log, sha2, rand, and tokio
   - Remove wildcard versions for tracing-subscriber and tokio
   - Add bin section to define the main binary and its path
 - Ensure all dependencies are explicitly versioned for better reproducibility

RandoomWalks committed:
Switch to asynchronous I/O and update time handling in Kademlia implementation

 - Replaced blocking I/O operations with asynchronous counterparts using Tokio, improving scalability and performance.
 - Changed time tracking within k-buckets from `SystemTime` to `Instant` to better support the async runtime.
 - Added serialization and deserialization for `Instant` to maintain compatibility with the existing data format.
 - Introduced an asynchronous main function with `#[tokio::main]` attribute to support async/await syntax.
 - Implemented a graceful shutdown mechanism using a message-passing channel.
 - Added periodic bucket refresh logic to maintain up-to-date network information.
 - Updated routing table and node handling to work within the async context.
 - General code cleanup and removal of unused imports and commented code.

hwrdtm commented on src/utils.rs:
```diff
@@ -0,0 +1,46 @@
+
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Debug)]
+pub struct NodeId([u8; 32]);
```
you might want to use hex encoding to more conveniently represent this for `Display` / `Debug`.

hwrdtm commented on src/utils.rs:
```diff
@@ -0,0 +1,46 @@
+        NodeId(random_bytes)
+    }
+
+    pub fn distance(&self, other: &NodeId) -> NodeId {
```
could be more useful as a "static method" as opposed to an "instance method"

hwrdtm commented on src/utils.rs:
```diff
@@ -0,0 +1,46 @@
+        }
+        NodeId(result)
+    }
+    pub fn from_slice(bytes: &[u8; 32]) -> Self {
```
nit: nice to implement using `From` / `Into` traits

hwrdtm commented on src/utils.rs:
```diff
@@ -0,0 +1,46 @@
+
+impl Ord for NodeId {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.cmp(&other.0)
```
is this not already the default implementation for ordering?

hwrdtm commented on src/message.rs:
```diff
@@ -0,0 +1,21 @@
+use std::net::SocketAddr;
+use serde::{Serialize, Deserialize};
+
+#[derive(Serialize, Deserialize, Debug)]
```
Do we want to specify any tagging styles here? https://serde.rs/enum-representations.html

hwrdtm commented on src/message.rs:
```diff
@@ -0,0 +1,21 @@
+    FindNode { target: NodeId },
+    FindValue { key: Vec<u8> },
+    NodesFound(Vec<(NodeId, SocketAddr)>),
+    ValueFound(Vec<u8>),
```
Maybe this might be more appropriate `FindValueRes(FindValueResult)`

hwrdtm commented on src/main.rs:
```diff
@@ -1,58 +1,45 @@
-    let addr2 = "127.0.0.1:8002".parse()?;
-    let mut node1 = KademliaNode::new(addr1).await;
-    let node2 = KademliaNode::new(addr2).await;
+    let addr: SocketAddr = "127.0.0.1:33334".parse().expect("Failed to parse socket address");
```
at some point we might want to introduce configurations and make this configurable.

hwrdtm commented on src/main.rs:
```diff
@@ -1,58 +1,45 @@
-        println!("Node1 successfully pinged Node2");
+    info!("Bootstrapping node...");
+    if let Err(e) = node.bootstrap().await {
+        error!("Bootstrap failed: {:?}", e);
```
unless you want to log here, `.map_err` might be cleaner. you could also log inside the `.map_err`

hwrdtm commented on src/node.rs:
```diff
@@ -0,0 +1,223 @@
+    pub async fn new(addr: SocketAddr) -> std::io::Result<(Self, mpsc::Sender<()>)> {
+        let id = NodeId::new();
+        let socket = UdpSocket::bind(addr).await?;
+        let (shutdown_sender, shutdown_receiver) = mpsc::channel(1);
```
semantically, if we're thinking that something _outside_ of this `node` mod would be sending to the channel to send shutdown signals, we might want to instantiate this channel _outside_ of this `node` mod, say in `src/main.rs`.

hwrdtm commented on src/node.rs:
```diff
@@ -0,0 +1,223 @@
+                        );
+                    }
+                }
+                Err(e) => error!("Invalid bootstrap address {:#?}: {:?}", bootstrap_addr, e),
```
should return the `Err` here

hwrdtm commented on src/node.rs:
```diff
@@ -0,0 +1,223 @@
+                        error!(
+                            "Failed to ping bootstrap node {:#?}: {:?}",
+                            bootstrap_addr, e
+                        );
```
should return the `Err` here

hwrdtm commented on src/node.rs:
```diff
@@ -0,0 +1,223 @@
+
+    pub async fn run(&mut self) -> std::io::Result<()> {
+        let mut buf = vec![0u8; 1024];
+        let mut refresh_interval = interval(Duration::from_secs(3600)); // Refresh every hour
```
You might want to set [MissedTickBehavior](https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html) 😉

hwrdtm commented on src/node.rs:
```diff
@@ -0,0 +1,223 @@
+        info!("Stored value for key: {:?}", hash);
+    }
+
+    fn hash_key(key: &[u8]) -> Vec<u8> {
```
more appropriate in utils

hwrdtm commented on src/node.rs:
```diff
@@ -0,0 +1,223 @@
+    id: NodeId,
+    addr: SocketAddr,
+    routing_table: RoutingTable,
+    storage: HashMap<Vec<u8>, Vec<u8>>,
```
you might want to use a `trait` for interfacing with the storage layer.

RandoomWalks committed:
Update Kademlia DHT Implementation with Enhanced Functionality

 This commit introduces a series of updates and enhancements to the Kademlia Distributed Hash Table (DHT) implementation, improving th
 overall functionality and robustness of the system from a user's perspective. Here are the key technical changes:

 - **VSCode Configuration**: Added a new `.vscode/launch.json` file to facilitate debugging of integration tests within the VSCode
 environment, specifically for Rust development.

 - **Struct Debugging**: Improved all structs by implementing the `fmt::Debug` trait, which allows for better logging and
 debugging of k-bucket entries.

 - **Library Module**: Created a new `src/lib.rs` file to define the library module, making it easier to use the Kademlia DHT as a
 dependency in test code.

 - **Unit Tests(All Pass)**: A new `src/unit_test.rs` file has been added, containing unit tests that cover various functionalities of the
 Kademlia DHT, such as node ID generation, bucket updating, and message handling.

 - **Integration Tests(Has Issues)**: Introduced a new `tests/integration_tests.rs` file that includes integration tests for node discovery, value
 storage and retrieval, and message handling between nodes.
